README.md
<img align="right" src="https://raw.github.com/cliffano/bagofrequest/master/avatar.jpg" alt="Avatar"/>

[![Build Status](https://img.shields.io/travis/cliffano/bagofrequest.svg)](http://travis-ci.org/cliffano/bagofrequest)
[![Dependencies Status](https://img.shields.io/david/cliffano/bagofrequest.svg)](http://david-dm.org/cliffano/bagofrequest)
[![Coverage Status](https://img.shields.io/coveralls/cliffano/bagofrequest.svg)](https://coveralls.io/r/cliffano/bagofrequest?branch=master)
[![Published Version](https://img.shields.io/npm/v/bagofrequest.svg)](http://www.npmjs.com/package/bagofrequest)
<br/>
[![npm Badge](https://nodei.co/npm/bagofrequest.png)](http://npmjs.org/package/bagofrequest)

Bag Of Request
--------------
Bag Of Request contains request utility functions.

bagofrequest#request

Send http request using [mikeal/request](http://github.com/mikeal/request), with the following additional features:

* status code-based response handlers registration
* wildcard status code support (e.g. 2xx, 50x)
* unexpected status code error handling with request body included in message
* request retry with increasing delay and maximum retries
* proxy setting based on env vars http_proxy, HTTP_PROXY, https_proxy, HTTPS_PROXY
* proxy exclusion for localhost and 127.0.0.1
* request timeout of 30 seconds
* follow redirects
* accepts self-signed SSL certificates
* all of the above defaults can be overridden

bagofrequest#proxy

Proxy retrieval based on URL and environment variables:

* if URL uses http, then sets proxy to http_proxy or HTTP_PROXY
* if URL uses https, then sets proxy to htps_proxy or HTTPS_PROXY, otherwise fallback to http_proxy or HTTP_PROXY
* if URL does not have a protocol, assume http protocol
* if URL is not provided, then set proxy to http_proxy or HTTP_PROXY, otherwise fallback to https_proxy or HTTPS_PROXY
* proxy will be ignored if host is on no_proxy or NO_PROXY when provided, otherwise ignore 127.0.0.1 and localhost

Installation
------------

    npm install bagofrequest

or as a dependency in package.json file:

    "dependencies": {
      "bagofrequest": "x.y.z"
    }

Usage
-----

    var bag = require('bagofrequest');

Request:

    // send http get request with query strings, timeout, and specified status code-based handlers
    bag.request('get', 'http://somehost', {
        queryStrings: {
          param1: 'value1',
          param2: 'value2'
        },
        timeout: 30000,
        handlers: {
          '2xx': function (result, cb) {
            cb(null, result.somedata);
          },
          404: function (result, cb) {
            cb(new Error('Item ' + result.itemId + ' cannot be found'));
          }
        }
      },
      function (err, result) {
        // response with unexpected status code (no registered handler) will be passed as error
      });

    // send http post request
    bag.request('post', 'http://somehost', {
        headers: {
          'content-type': 'application/json'
        },
        json: {
          prop1: 'value1',
          prop2: 'value2'
        }
      },
      function (err, result) {
      });

    // send request with options to override any bagofrequest defaults (will be passed to mikeal/request)
    bag.request('get', 'http://somehost', {
        requestOpts: {
          foo: 'bar'
        }
      },
      function (err, result) {
      });

    // send request with retry settings
    bag.request('get', 'http://somehost', {
        retry: {
          errorCodes: true, // retry on any error
          statusCodes: [404, 503], // retry when response status code is 404 or 503
          scale: 0.5, // increase delay by half on each retry
          delay: 1000, // wait 1 second before retrying
          maxRetries: 10 // only retry 10 times at most
        }
      },
      function (err, result) {
      });

    // send request with custom proxy
    bag.request('get', 'http://somehost', {
        proxy: 'http://user:pass@someproxy:1234'
      },
      function (err, result) {
      });

Proxy:

    // get proxy based on URL protocol
    // will return undefined when host is localhost or 127.0.0.1
    var proxy = bag.proxy('https://somehost');

    // get proxy with custom proxy exclusion
    var proxy = bag.proxy('http://somelocalhost', {
      noProxyHosts: ['somelocalhost']
    });

Colophon
--------

[Developer's Guide](http://cliffano.github.io/developers_guide.html#nodejs)

Build reports:

* [Code complexity report](http://cliffano.github.io/bagofrequest/complexity/plato/index.html)
* [Unit tests report](http://cliffano.github.io/bagofrequest/test/buster.out)
* [Test coverage report](http://cliffano.github.io/bagofrequest/coverage/buster-istanbul/lcov-report/lib/index.html)
* [Integration tests report](http://cliffano.github.io/bagofrequest/test-integration/buster.out)
* [API Documentation](http://cliffano.github.io/bagofrequest/doc/dox-foundation/index.html)


CHANGELOG.md
### 1.0.1
*

### 1.0.0
*

### 0.1.5
*

### 0.1.4
* Add noproxy environment variable support

### 0.1.3
* Set min node engine to v4.0.0

### 0.1.2
* Add agentOptions param passing [Benedikt Arnold](https://github.com/benediktarnold)

### 0.1.1
* Add build reports to readme

### 0.1.0
* Set min node engine to v0.10.0

### 0.0.6
* Change test to use buster-node + referee
* Set min node engine to v0.8.0

### 0.0.5
* Add custom error handler and error codes matching to retry support [Daniel Brain](https://github.com/dbrain)

### 0.0.4
* Add request retry support [Daniel Brain](https://github.com/dbrain)

### 0.0.3
* Add no-proxy hosts support to proxy
* Fix proxy to fallback to HTTP_PROXY when http_proxy doesn't exist

### 0.0.2
* Set rejectUnauthorized default to false
* Modify default timeout to 30 seconds

### 0.0.1
* Initial version, extracted from cliffano/bagofholding


complexity/plato/report.js
__report = {"summary":{"total":{"jshint":0,"sloc":208,"maintainability":64.64608874146919},"average":{"sloc":208,"maintainability":"64.65","jshint":"0.00"}},"reports":[{"info":{"file":"lib/bagofrequest.js","fileShort":"lib/bagofrequest.js","fileSafe":"lib_bagofrequest_js","link":"files/lib_bagofrequest_js/index.html"},"jshint":{"messages":0},"complexity":{"aggregate":{"sloc":{"logical":113,"physical":208},"cyclomatic":39,"halstead":{"operators":{"distinct":26,"total":324,"identifiers":["__stripped__"]},"operands":{"distinct":124,"total":414,"identifiers":["__stripped__"]},"length":738,"vocabulary":150,"difficulty":43.403225806451616,"volume":5334.8681935859595,"effort":231550.48885386804,"bugs":1.7782893978619865,"time":12863.916047437113},"params":24,"line":1,"cyclomaticDensity":34.51327433628318,"complexity":{"cyclomatic":39,"sloc":{"logical":113,"physical":208},"halstead":{"operators":{"distinct":26,"total":324,"identifiers":["__stripped__"]},"operands":{"distinct":124,"total":414,"identifiers":["__stripped__"]},"length":738,"vocabulary":150,"difficulty":43.403225806451616,"volume":5334.8681935859595,"effort":231550.48885386804,"bugs":1.7782893978619865,"time":12863.916047437113}}},"module":"lib/bagofrequest.js","maintainability":64.64608874146919}}]}

lib/bagofrequest.js
var request = require('request'),
  _url = require('url'),
  async = require('async');

/**
 * Sends a HTTP request to a specified URL with optional proxy, query strings, and handlers.
 * Convenient handling of request error and unexpected status code.
 * GET redirection is handled by default by request module.
 * Non-GET redirection is handled by params followAllRedirects: true.
 *
 * @param {String} method: http method
 * @param {String} url: URL without query string
 * @param {Object} opts: optional
 *   - proxy: proxy server URL with format http://user:pass@host:port
 *   - noProxyHosts: an array of host names which will ignore any proxy settings, defaults to: localhost, 127.0.0.1
 *   - queryStrings: object containing URL query strings with format { name: value }
 *   - handlers: response handlers with format { error: function(err, result, cb), statuscode: function(result, cb) }
 *   - headers: object containing http headers, exact map to mikeal/request headers opt
 *   - body, form, json, multipart, agentOptions: request payload, exact map to mikeal/request module body, form, json, multipart, agentOptions opts
 *   - retry: request retry configuration, false for disabled, true for sensible defaults, otherwise
 *            { errorCodes: ['ENOTFOUND'] || true/false, statusCodes: [ 503 ], scale: 0.5, delay: 1000, maxRetries: 10 }
 *   - timeout: request timeout in milliseconds, default to 30000 milliseconds
 *   - requestOpts: mikeal/request module opts, these opts will override any params set by http#req
 * @param {Function} cb: standard cb(err, result) callback
 */
function req(method, url, opts, cb) {
  opts.handlers = opts.handlers || {};
  opts.retry = opts.retry || false;

  const TIMEOUT = 30000;
  // init default settings with follow redirections on all methods, and 30 seconds timeout
  var params = {
      url: url,
      followAllRedirects: true,
      rejectUnauthorized: false,
      timeout: opts.timeout || TIMEOUT
    },
    envProxy = proxy(url);

  // set proxy setting based on environment variable
  if (!_proxyIgnored(url, opts)) {
    if (opts.proxy) {
      params.proxy = opts.proxy;
    } else if (envProxy) {
      params.proxy = envProxy;
    }
  }

  // just a more readable opt name for query strings
  if (opts.queryStrings) {
    params.qs = opts.queryStrings;
  }

  // headers and payload handling
  ['headers', 'body', 'form', 'json', 'multipart','agentOptions'].forEach(function (opt) {
    if (opts[opt]) {
      params[opt] = opts[opt];
    }
  });

  // override defaults with mikeal/request-specific opts
  if (opts.requestOpts) {
    Object.keys(opts.requestOpts).forEach(function (opt) {
      params[opt] = opts.requestOpts[opt];
    });
  }

  function _requestMethod(method) {
    var requestMethod = method.toLowerCase();
    if (requestMethod === 'delete') {
      requestMethod = 'del';
    }
    return requestMethod;
  }

  _execRequest(opts, _requestMethod(method), params, function (err, result) {
    if (err) {
      if (opts.handlers.error) {
        opts.handlers.error(err, result, cb);
      } else {
        cb(err);
      }
    } else {
      var wildcardMatch = _wildcardMatch(result.statusCode, Object.keys(opts.handlers));
      if (opts.handlers[result.statusCode]) {
        opts.handlers[result.statusCode](result, cb);
      } else if (wildcardMatch) {
        opts.handlers[wildcardMatch](result, cb);
      } else {
        cb(new Error('Unexpected status code: ' + result.statusCode + '\nResponse body:\n' + result.body));
      }
    }
  });
}

function _execRequest(opts, method, params, cb) {
  const DEFAULTS = {
    errorCodes: [ 'ECONNABORTED', 'ECONNREFUSED', 'ECONNRESET', 'EHOSTUNREACH', 'EIO', 'ENETDOWN', 'ENETRESET', 'ENETUNREACH', 'ETIMEDOUT' ],
    statusCodes: [ 408, 502, 503, 504 ],
    scale: 0.5,
    delay: 500,
    maxRetries: 10
  };

  var tryCount = 0,
    currDelay, lastResult, retryable,
    retryOpts = (opts.retry === true) ?
      DEFAULTS :
      {
        errorCodes: opts.retry.errorCodes || DEFAULTS.errorCodes,
        statusCodes: opts.retry.statusCodes || DEFAULTS.statusCodes,
        scale: (opts.retry.scale !== undefined) ? opts.retry.scale : DEFAULTS.scale,
        delay: (opts.retry.delay !== undefined) ? opts.retry.delay : DEFAULTS.delay,
        maxRetries: (opts.retry.maxRetries !== undefined) ? opts.retry.maxRetries : DEFAULTS.maxRetries
      };

  currDelay = retryOpts.delay;

  function does(cb) {
    request[method](params, function (err, result) {

      var retryableError = err && (retryOpts.errorCodes === true || (Array.isArray(retryOpts.errorCodes) && retryOpts.errorCodes.indexOf(err.code) !== -1)),
        retryableResult = result && _wildcardMatch(result.statusCode, retryOpts.statusCodes) !== undefined;

      tryCount++;
      lastResult = result;

      retryable = opts.retry && tryCount <= retryOpts.maxRetries && (retryableError || retryableResult);

      if (retryable) {
        setTimeout(cb, currDelay); // retry after delay
        currDelay += currDelay * retryOpts.scale;
      } else {
        cb(err); // no retry needed, finish off
      }
    });
  }

  function check() {
    return retryable;
  }

  function done(err) {
    var result = lastResult || {};
    result._retry = {
      retryCount: tryCount - 1,
      retryLimitHit: (tryCount - 1 === retryOpts.maxRetries)
    };
    cb(err, result);
  }

  async.doWhilst(does, check, done);
}

// check for any wildcard match (e.g. 201 matches 2xx)
function _wildcardMatch(statusCode, wildcards) {
  var match;
  wildcards.forEach(function (wildcard) {
    var regex = new RegExp(wildcard.toString().replace(/x/g, '.'));
    if (!match && statusCode.toString().match(regex)) {
      match = wildcard;
    }
  });
  return match;
}

/**
 * Determines proxy value based on URL and process environment variable (http_proxy, https_proxy).
 * This allows library clients to control which proxy to use by setting environment variable.
 * - if url starts with http, use http_proxy when available
 * - if url starts with https, use https_proxy when available, otherwise fallback to http_proxy
 * - if url does not have protocol, assume http protocol
 * - if url is not specified, http_proxy takes precedence over https_proxy
 *
 * @param {String} url: URL used to determine which proxy environment variable to use
 * @param {Object} opts: optional
 *   - noProxyHosts: an array of host names which will ignore any proxy settings, defaults to: localhost, 127.0.0.1
 * @return proxy
 */
function proxy(url, opts) {
  opts = opts || {};

  var _proxy;

  if (!url) {
    _proxy = process.env.http_proxy || process.env.HTTP_PROXY || process.env.https_proxy || process.env.HTTPS_PROXY;
  } else if (!_proxyIgnored(url, opts)) {
    if (!url.match(/^https?:\/\//)) {
      url += 'http://' + url;
    }

    if (url.match(/^https:\/\//)) {
      _proxy = process.env.https_proxy || process.env.HTTPS_PROXY || process.env.http_proxy || process.env.HTTP_PROXY;
    } else {
      _proxy = process.env.http_proxy || process.env.HTTP_PROXY;
    }
  }

  return _proxy;
}

function _proxyIgnored(url, opts) {
  const NO_PROXY_HOSTS = ['localhost', '127.0.0.1'];
  var noProxyHosts = [];
  if (process.env.no_proxy) {
    noProxyHosts = noProxyHosts.concat(process.env.no_proxy.split(','));
  }
  if (process.env.NO_PROXY) {
    noProxyHosts = noProxyHosts.concat(process.env.NO_PROXY.split(','));
  }
  noProxyHosts = noProxyHosts.concat(NO_PROXY_HOSTS);
  return (opts.noProxyHosts || noProxyHosts).indexOf(_url.parse(url).hostname) !== -1;
}

exports.request = req;
exports.proxy = proxy;
